File: client-lib/server/src/main/java/dev/boiarshinov/breakingbad/clientlib/server/CrystalController.java
package dev.boiarshinov.breakingbad.clientlib.server;

import dev.boiarshinov.breakingbad.clientlib.datamodel.CrystalUpdateRequestDto;
import dev.boiarshinov.breakingbad.clientlib.datamodel.CrystalUpdatedDto;
import dev.boiarshinov.breakingbad.clientlib.datamodel.ResponseDto;
import org.springframework.core.MethodParameter;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.annotation.*;

import java.util.Map;
import java.util.UUID;

@RestController
@RequestMapping("/v1/crystals")
public class CrystalController {
    @PutMapping("/{id}")
    public ResponseEntity<ResponseDto<CrystalUpdatedDto>> updateCrystal(
            @RequestHeader("X-CUSTOM-TOKEN") String token,
            @PathVariable("id") String id,
            @RequestBody CrystalUpdateRequestDto request
    ) {
        var crystalUpdatedDto = new CrystalUpdatedDto(UUID.randomUUID().toString(), request.type(), request.price());
        return ResponseEntity.ok(new ResponseDto<>("S000", "Success", crystalUpdatedDto));
    }

    @ExceptionHandler(MethodArgumentNotValidException.class)
    private ResponseEntity<ResponseDto> handleInvalidArguments(MethodArgumentNotValidException ex) {
        Map<String, MethodParameter> responseMap = Map.of("param", ex.getParameter());

        return ResponseEntity.badRequest()
                .body(new ResponseDto("E802", "Invalid Request Params", responseMap));
    }
}

File: contract-testing/spring-cloud-contract/heisenberg-scc-provider/src/main/kotlin/io/github/artemptushkin/breakingbad/scc/CrystalController.kt
package io.github.artemptushkin.breakingbad.scc

import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController
import java.math.BigDecimal

@RestController
@RequestMapping("/v1/crystals")
class CrystalController {
    @GetMapping
    operator fun get(@RequestParam amount: Int): CrystalsResponse {
        val price = BigDecimal("10.0")
        return CrystalsResponse(
            Crystal(1L, price.multiply(BigDecimal.valueOf(amount.toLong())))
        )
    }
}